README.md
# Discord Ticket System by AstroGD
A ticket system for Discord to reach the users questions more easily and never loose a message in chat.

Simply edit command names and language via language and config files.

If you have any question feel free to contact me via mail to support@astrogd.eu or send a dm to AstroGD#3416

## Features
Getting in touch with the team on a server with many members can be very difficult and frustrating. Your question or the teams answer might get lost in the chat and you need to scroll and search if you want to re-read the messages. Thats why the ticketsystem is for you!
If you have questions you can start a ticket and the team will be notified about your issue. But there's more: You'll even get a DM from the bot if your case updates so that you also won't miss an update ever again.

So what are are you waiting for? Download ticketsystem now and make reaching you and your team easier than ever before!

## Screenshots
![Create a ticket](https://imgur.com/Ox0nlfs.png)

_A ticket was created_

![Answer to a ticket](https://imgur.com/480fpzG.png)

_A ticket has been answered_

## How to use
- If you want to install this project go to [How to install](https://github.com/AstroGD/ticketsystem/wiki/How-to-install)
- For help in setting up your config go to [Set up your config](https://github.com/AstroGD/ticketsystem/wiki/Set-up-your-config)
- Looking for a list of all available commands? Go to [available commands](https://github.com/AstroGD/ticketsystem/wiki/Available-commands)
- Still having an issue or a question? [Reach out to me](https://www.astrogd.eu/en/kontakt)

## I noticed a bug what can I do?
- Please [create a bug report](https://github.com/AstroGD/ticketsystem/issues/new/choose) and tell me as precise as possible what happened, how to reproduce the bug and which Node version and version of the ticketsystem you are running


changelog.md
# TBA
#### Released on TBA

## Updates
:arrow_up: Update follow-redirects to 1.14.0

# Version 2.0.5
#### Released on 10.04.2021

## Bugfixes
:bug: Fixed a bug where the bot would crash if an invalid guild id was provided in the config

## Dependency Updates
:arrow_up: Update discord.js to 12.5.3
:arrow_up: Update follow-redirects to 1.13.3

# Version 2.0.4
#### Released on 03.02.2021

## Dependency updates
:arrow_up: Update discord.js to 12.5.1
:arrow_up: Update follow-redirects to 1.13.2
:arrow_up: Update fs-extra to 9.1.0

# Version 2.0.3
#### Released 01.10.2020 (DMY)

## Bugfixes
- :bug: Fixed a bug where no Notification is being sent after a user answers a ticket
- :bug: Fixed a bug where /ticketteam would not send an error if a user has insufficient permissions
- :bug: Fixed a bug where sending /ticketteam without an action would throw no error for authorized users
- :bug: Fixed embed colour for error message if an user with insufficient permissions tries to use /ticketteam

## Dependency updates
- :arrow_up: Upgraded dependencies to newest version

## Removed Features
- :fire: Removed legacy check for me to be able to use /ticketteam without having the team-role

## Language
- :globe_with_meridians: Fixed a typo for English language pack

## UI and Style
- :lipstick: Log output regarding version checks are now colourized. If no update is available the output will be coloured green. If there is an update available or an error occurred while checking for a new version the output will be red.

## Documentation
- :memo: Added Changelog

request.js
module.exports = function (opt) {
    return new Promise((resolve, reject) => {
        const https = require('follow-redirects').https;

        var options = opt || {
            'method': 'GET',
            'hostname': 'software.astrogd.eu',
            'path': '/ticketsystem/version.json',
            'headers': {},
            'maxRedirects': 20
        };

        var req = https.request(options, function (res) {
            var chunks = [];

            res.on("data", function (chunk) {
                chunks.push(chunk);
            });

            res.on("end", function () {
                var body = Buffer.concat(chunks);
                resolve(body);
            });

            res.on("error", function (error) {
                reject(error);
            });
        });

        req.end();
    });
}

failsafe.js
/*Failsafe System Version 1.1.6
(CC-BY-NC-ND-4.0) AstroGD - www.astrogd.de
Edited by Cherob - www.cherob.eu */

//Required Modules and Files
const cp = require("child_process");
const package = require(`${__dirname}/package.json`);

let title = "   ______      _  _   _____ \n"
  + "   |  ____|    (_)| | / ____| \n"
  + "   | |__  __ _  _ | || (___    __ _ __   __ ___ \n"
  + "   |  __|/ _` || || | \\___ \\  / _` |\\ \\ / // _ \\ \n"
  + "   | |  | (_| || || | ____) || (_| | \\ V /|  __/ \n"
  + "   |_|   \\__,_||_||_||_____/  \\__,_|  \\_/  \\___|"
var filedirectory = "/main.js";
var version = "1.1.5";

let unixtime = ('[' + Math.round((new Date()).getTime() / 1000) + ']');

function start() {
  console.log(`\n ${title}  - V${version} -\n`);
  let programm = cp.fork(`${__dirname}${filedirectory}`);
  programm.on('exit', (code, signal) => {
    switch (code) {
      case 28:
        console.log(`\nAn unknown Error occurred. Shutting down.`
          + `\nThe Programm is being closed in 3 Seconds`);
        setTimeout(function () { process.exit(); }, 3000);
        break;
      case 29:
        console.log(`\nSystem thread encountered an error.`
          + `\nThe Programm is being restarted in 5 Seconds`);
        setTimeout(function () { start(); }, 5000);
        break;
      default:
        console.log(`\nSystem thread crashed unexpectedly`
          + `\nCode: ${code}`
          + `\nSignal: ${signal}`
          + `\nThe Programm is being restarted in 10 Seconds`);
        setTimeout(function () { start(); }, 10000);
        break;
    }
  });
  console.log(unixtime, `Starting "${package.name}.js"`);
  console.log(unixtime, "OUTPUT PROGRAMM");
}

//startup after loading
start();


makeEmbed.js
module.exports = function (client) {
    return function (message, color, title, description, fields) {
        const Discord = require("discord.js");

        var embed = new Discord.MessageEmbed();
        embed.setColor(color)
            .setFooter("Ticket System by AstroGD", client.user.avatarURL()) //Getting my name out in the world is the only payment I get for releasing free code so I am very thankful if you leave this as it is.
            .setTimestamp(new Date())
            .setTitle(title)
            .setDescription(description);

        if (message) {
            embed.setAuthor(message.author.username, message.author.avatarURL());
        }

        if (fields) {
            for (var i = 0; i < fields.length; i++) {
                if (!fields[i].inline) fields[i].inline = false;
                if (!fields[i].empty) {
                    embed.addField(fields[i].title, fields[i].value, fields[i].inline);
                } else {
                    embed.addField('\u200b', '\u200b', fields[i].inline);
                }
            }
        }
        return embed;
    }
}

commands/hilfe.js
module.exports = function (log, makeEmbed, language, config, languageconfig) {
    return function (message) {
        var embed = makeEmbed(message, "0094FF", `ℹ️ ${language.help}`, language.information_to_the_use_of_the_ticket_system, [{
            title: `${config.prefix}${languageconfig.ticketcommand.main} ${languageconfig.ticketcommand.create} <${language.message}>`,
            value: language.starts_a_new_ticket
        }, {
            title: `${config.prefix}${languageconfig.ticketcommand.main} ${languageconfig.ticketcommand.info} <${language.ticketid}>`,
            value: language.information_about_the_selected_ticket
        }, {
            title: `${config.prefix}${languageconfig.ticketcommand.main} ${languageconfig.ticketcommand.answer} <${language.ticketid}> <${language.message}>`,
            value: language.sends_an_answer_to_a_ticket
        }, {
            title: `${config.prefix}${languageconfig.ticketcommand.main} ${languageconfig.ticketcommand.close} <${language.ticketid}>`,
            value: language.closes_a_ticket
        }]);
        message.channel.send("", {
            embed: embed
        }).catch(e => {
            log(`An error occurred while sending a message to ${message.channel.name}: ${e.stack}`, 1);
        }).then(msg => {
            msg.delete({timeout:30000}).catch(e => {});
        });

        var guild = message.guild;
        var role = guild.roles.cache.get(config.team_role_id);

        if (!role) {
            log(`Teamroleid invalid. Please check the config.json`, 1);
            message.channel.send("", {
                    embed: makeEmbed(message, "#00FF21", `❗ ${language.an_error_occurred}`, language.the_team_role_stored_in_the_bots_config_is_invalid_please_contact_an_administrator)
                })
                .catch(e => {
                    log(`An error occured while sending a message to ${message.channel.name}: ${e.stack}`, 1);
                });
            return;
        }

        if (message.member.roles.cache.get(role.id)) {
            message.author.send("", {
                    embed: makeEmbed(message, "#0094FF", `ℹ️ ${language.help}`, language.information_to_the_use_of_the_ticket_system + " | " + language.for_staff_members, [{
                        title: `${config.prefix}${languageconfig.ticketteamcommand.main} ${languageconfig.ticketteamcommand.info} <${language.ticketid}>`,
                        value: language.information_about_the_selected_ticket
                    }, {
                        title: `${config.prefix}${languageconfig.ticketteamcommand.main} ${languageconfig.ticketteamcommand.answer} <${language.ticketid}> <${language.message}>`,
                        value: language.sends_an_answer_to_a_ticket
                    }, {
                        title: `${config.prefix}${languageconfig.ticketteamcommand.main} ${languageconfig.ticketteamcommand.close} <${language.ticketid}>`,
                        value: language.closes_a_ticket
                    }, {
                        title: `${config.prefix}${languageconfig.ticketteamcommand.main} ${languageconfig.ticketteamcommand.reopen} <${language.ticketid}>`,
                        value: language.reopens_a_ticket
                    }, {
                        title: `${config.prefix}${languageconfig.ticketteamcommand.main} ${languageconfig.ticketteamcommand.list}`,
                        value: language.shows_a_list_of_all_open_tickets
                    }, {
                        title: `${config.prefix}${languageconfig.ticketteamcommand.main} ${languageconfig.ticketteamcommand.user} <${language.userid_or_mention}>`,
                        value: language.shows_a_list_of_tickets_created_by_a_specific_user
                    }])
                })
                .catch(e => {
                    log(`An error occured while sending a message to ${message.author.tag}: ${e.stack}`, 1);
                });
            return;
        }
    }
}

main.js
/**
 * Ticketsystem by AstroGD
 * @author AstroGD (https://www.astrogd.eu)
 * @version 2.0.3
 * @since 15. Mai 2018
 * @link https://www.astrogd.eu
 * @license AGPL-3.0
 */

console.log(`[I] Loader: Loading Code`);

//Requirements
const Discord = require("discord.js");
const fs = require("fs-extra");
const path = require("path");

//Variables
const startupdate = new Date();
const timestamp = `${startupdate.getFullYear()}-${startupdate.getMonth() + 1}-${startupdate.getDate()}-${startupdate.getHours()}-${startupdate.getMinutes()}-${startupdate.getSeconds()}`;
const client = new Discord.Client();
const pk = require(path.join(__dirname, "/package.json"));
var language, languageconfig, commands;

//Modules
const request = require(path.join(__dirname, "/request.js"));
const makeEmbed = require(path.join(__dirname, "/makeEmbed.js"))(client);

//Path Ensurances
fs.ensureDirSync(path.join(__dirname, "/log/"));
fs.ensureDirSync(path.join(__dirname, "/tickets/"));
fs.ensureDirSync(path.join(__dirname, "/lang"));

//Files
var config = require(`${__dirname}/config.json`);
var stream = fs.createWriteStream(`${__dirname}/log/${timestamp}.log`);

//Functions
function log(msg, status = "INFO", consoletype = 1) {
    let dateCurrent = new Date();
    const timestampCurrent = `${dateCurrent.getFullYear()}-${dateCurrent.getMonth() + 1}-${dateCurrent.getDay()}-${dateCurrent.getHours()}-${dateCurrent.getMinutes()}-${dateCurrent.getSeconds()}`;

    switch (status) {
        case 1: //ERROR
            status = "ERROR";
            break;
        case 2: //WARN
            status = "WARN";
            break;
        case 3: //DEBUG
            status = "DEBUG";
            break;
        default: //INFO
            status = "INFO";
            break;
    }

    if (!consoletype || (consoletype != 1 && consoletype != 2 && consoletype != 3)) consoletype = 1;

    let logCurrent = `[${timestampCurrent}] ${status} [Ticketsystem] ${msg.stack ? `An error occurred:\n${msg.stack}` : msg}`;
    if (consoletype == 1 || consoletype == 3) stream.write(logCurrent + '\n');
    if (consoletype == 1 || consoletype == 2) console.log(logCurrent);
}

async function checkVersion() {
    let versionInfo = null;
    try {
        versionInfo = await request({
            'method': 'GET',
            'hostname': 'software.astrogd.eu',
            'path': '/ticketsystem/version.json',
            'headers': {},
            'maxRedirects': 20
        });
    } catch (error) {
        return log(`\x1b[31mVersion couldn't be checked! Make sure to allow network access to software.astrogd.eu\x1b[0m`, 1);
    }

    if (!versionInfo) return log(`\x1b[31mVersion couldn't be checked! Make sure to allow network access to software.astrogd.eu\x1b[0m`, 1);
    versionInfo = JSON.parse(versionInfo.toString());

    const version = pk.version.split(".");
    const latest = versionInfo.version.split(".");

    if (version[0] < latest[0] || version[0] == latest[0] && version[1] < latest[1] || version[0] == latest[0] && version[1] == latest[1] && version[2] < latest[2]) {
        return log(`\x1b[31mA new version of the ticketsystem is available (${pk.version} --> ${versionInfo.version}). Make sure to visit ${versionInfo.url} and download the newest version. If you need help look in the repositorys wiki under "how to update" or contact AstroGD (https://www.astrogd.eu/en/kontakt)\x1b[0m`, 2);
    }

    log(`\x1b[32mNewest version installed!\x1b[0m`, 3);
};

//Additional Code
process.on("unhandledRejection", (reason) => {
    log(`An unhandled Promise Rejection occurred: ${reason}`, 1);
});

process.on('uncaughtException', (err) => {
    try {
        log(`An uncaught Exception took Place: ${err.stack}\n\nBot is force destroying...`, 1);
    } catch (e) {

    }
    try {
        client.destroy().then(() => {
            process.exit();
        });
    } catch (e) {
        process.exit();
    }
    return;
});

//Startup Function
async function startup() {
    log(`Initialized!`);
    log("Starting up - Loading Bots language");

    if (!fs.existsSync(`${__dirname}/lang/${config.language}.json`)) {
        log(`Language File not found. Please download missing files or contact the developer.`, 1);
        process.exit(28);
    } else {
        language = JSON.parse(fs.readFileSync(`${__dirname}/lang/${config.language}.json`, "utf-8")).translations;
        languageconfig = JSON.parse(fs.readFileSync(`${__dirname}/lang/${config.language}.json`, "utf-8")).settings;
    }

    log("Done - Logging in");

    await checkVersion();
    setInterval(function () {
        checkVersion();
    }, 3600000); //Check version every hour

    client.login(config.token);

    client.on("ready", () => {
        try {
            log(`Online. Connected as ${client.user.tag} (${client.user.id}) - Checking Botuser`);
            if (!client.user.bot) {
                log(`Not a Bot User. This is against the Discord ToS. Program will be terminated. Please change the token in the config.`, 1);
                client.destroy()
                    .then(() => {
                        process.exit();
                    });
                return;
            }

            log("Is a bot user - Checking Guild");
            let guild = client.guilds.cache.get(config.server_id);
            if (!guild) {
                log(`Guild not found. Please edit Config.json. Program will be terminated.`, 1);
                client.destroy();
                process.exit();
            }

            log(`Guild found. Listening to ${guild.name} (${guild.id}) - Loading commands`);

            client.user.setPresence({
                status: "online",
                activity: {
                    name: config.custom_status || `${language.ticket_system} | ${config.prefix}${languageconfig.helpcommand.main}`,
                    type: "PLAYING"
                }
            });

            //Commands
            const ticketCommand = require(path.join(__dirname, "/commands/ticket.js"))(log, makeEmbed, language, config, languageconfig, client);
            const ticketteamCommand = require(path.join(__dirname, "/commands/ticketteam.js"))(log, makeEmbed, language, config, languageconfig, client);
            const hilfeCommand = require(path.join(__dirname, "/commands/hilfe.js"))(log, makeEmbed, language, config, languageconfig);
            commands = {
                ticket: ticketCommand,
                ticketteam: ticketteamCommand,
                hilfe: hilfeCommand
            }

            log(`Done - Listening for commands.`);
        } catch (e) {
            log(`An error occurred while running Ready Event Actions: ${e.stack}`, 1);
        }
    });

    client.on("message", async (message) => {
        try {
            if (message.author.bot) return;
            if (!message.guild) {
                log(`Got a DM from ${message.author.tag} - Sorry but thats unwanted`);
                message.author.send("", {
                        embed: makeEmbed(message, "#FF0000", `🚫 - ${language.oops}! ${language.this_bot_is_not_usable_via_dm}`, language.please_enter_your_command_on_our_server)
                    })
                    .catch(e => {
                        log(`An error occurred while sending a DM to ${message.author.tag}: ${e.stack}`, 2);
                        return;
                    });
                return;
            }
            if (!message.content.startsWith(config.prefix) || message.guild.id != config.server_id) return;

            log(`Received commandlike message from ${message.author.tag} - Checking for Commands`);

            const cmd = message.content.split(config.prefix).slice(1).join(config.prefix).split(" ")[0].toLowerCase();

            if (cmd == languageconfig.ticketcommand.main.toLowerCase()) {
                message.content = `${config.prefix}ticket ${message.content.split(" ").slice(1).join(" ")}`;
            } else if (cmd == languageconfig.ticketteamcommand.main.toLowerCase()) {
                message.content = `${config.prefix}ticketteam ${message.content.split(" ").slice(1).join(" ")}`;
            } else if (cmd == languageconfig.helpcommand.main.toLowerCase()) {
                message.content = `${config.prefix}hilfe ${message.content.split(" ").slice(1).join(" ")}`;
            }

            if (!commands[message.content.split(config.prefix).slice(1).join(config.prefix).split(" ")[0].toLowerCase()]) {
                log(`No Command found :(`);
                if (config.errorOnCommandNotFound) {
                    message.channel.send("", {
                            embed: makeEmbed(message, "#FF0000", `⁉️ ${language.oops}`, language.this_command_does_not_exist_please_use_the_help_command_to_get_a_list_of_all_commands)
                        })
                        .catch(e => {
                            log(`An error occurred while sending a message to ${message.channel.name}: ${e.stack}`, 1);
                            return;
                        })
                        .then(msg => {
                            msg.delete({timeout:10000})
                                .catch(e => {});
                        });
                }
                return;
            } else {                
                log(`Command found :) | ${message.content.split(" ")[0].toLowerCase()}`);
                commands[message.content.split(config.prefix).slice(1).join(config.prefix).split(" ")[0].toLowerCase()](message);
                message.delete({timeout:0}).catch(e => {});
                return;
            }
        } catch (e) {
            log(`An error occurred while running Message Event Actions: ${e.stack}`, 1);
        }
    });
}


//Startup
console.log(`[I] Loader: Code loaded. Starting Function - Activating Log Function`);
startup();